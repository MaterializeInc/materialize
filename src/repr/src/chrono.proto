// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

syntax = "proto3";

package mz_repr.chrono;

message ProtoTz {
    string name = 1;
}

message ProtoNaiveDate {
   int32 year = 1;
   uint32 ordinal = 2;
}

message ProtoNaiveTime {
    uint32 secs = 1;
    uint32 frac = 2;
}

message ProtoNaiveDateTime {
    ProtoNaiveDate date = 1;
    ProtoNaiveTime time = 2;
}

message ProtoFixedOffset {
    int32 local_minus_utc = 1;
}
