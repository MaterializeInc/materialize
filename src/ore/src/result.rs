// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Apache license, Version 2.0

//! Result utilities.

/// Extension methods for [`std::result::Result`].
pub trait ResultExt<T, E> {
    /// Applies [`Into::into`] to a contained [`Err`] value, leaving an [`Ok`]
    /// value untouched.
    fn err_into<E2>(self) -> Result<T, E2>
    where
        E: Into<E2>;
}

impl<T, E> ResultExt<T, E> for Result<T, E> {
    fn err_into<E2>(self) -> Result<T, E2>
    where
        E: Into<E2>,
    {
        self.map_err(|e| e.into())
    }
}
