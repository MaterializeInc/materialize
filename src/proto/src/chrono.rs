// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Custom [`proptest::strategy::Strategy`] implementations and Protobuf types
//! for the [`chrono`] fields used in the codebase.
//!
//! See the [`proptest`] docs[^1] for an example.
//!
//! [^1]: <https://altsysrq.github.io/proptest-book/proptest-derive/modifiers.html#strategy>

use chrono::{
    DateTime, Datelike, Duration, FixedOffset, NaiveDate, NaiveDateTime, NaiveTime, Timelike, Utc,
};
use chrono_tz::{TZ_VARIANTS, Tz};
use proptest::prelude::Strategy;

use crate::{RustType, TryFromProtoError};

include!(concat!(env!("OUT_DIR"), "/mz_proto.chrono.rs"));

impl RustType<ProtoNaiveDate> for NaiveDate {
    fn into_proto(&self) -> ProtoNaiveDate {
        ProtoNaiveDate {
            year: self.year(),
            ordinal: self.ordinal(),
        }
    }

    fn from_proto(proto: ProtoNaiveDate) -> Result<Self, TryFromProtoError> {
        NaiveDate::from_yo_opt(proto.year, proto.ordinal).ok_or_else(|| {
            TryFromProtoError::DateConversionError(format!(
                "NaiveDate::from_yo_opt({},{}) failed",
                proto.year, proto.ordinal
            ))
        })
    }
}

impl RustType<ProtoNaiveTime> for NaiveTime {
    fn into_proto(&self) -> ProtoNaiveTime {
        ProtoNaiveTime {
            secs: self.num_seconds_from_midnight(),
            frac: self.nanosecond(),
        }
    }

    fn from_proto(proto: ProtoNaiveTime) -> Result<Self, TryFromProtoError> {
        NaiveTime::from_num_seconds_from_midnight_opt(proto.secs, proto.frac).ok_or_else(|| {
            TryFromProtoError::DateConversionError(format!(
                "NaiveTime::from_num_seconds_from_midnight_opt({},{}) failed",
                proto.secs, proto.frac
            ))
        })
    }
}

impl RustType<ProtoNaiveDateTime> for NaiveDateTime {
    fn into_proto(&self) -> ProtoNaiveDateTime {
        ProtoNaiveDateTime {
            year: self.year(),
            ordinal: self.ordinal(),
            secs: self.num_seconds_from_midnight(),
            frac: self.nanosecond(),
        }
    }

    fn from_proto(proto: ProtoNaiveDateTime) -> Result<Self, TryFromProtoError> {
        let date = NaiveDate::from_yo_opt(proto.year, proto.ordinal).ok_or_else(|| {
            TryFromProtoError::DateConversionError(format!(
                "NaiveDate::from_yo_opt({},{}) failed",
                proto.year, proto.ordinal
            ))
        })?;

        let time = NaiveTime::from_num_seconds_from_midnight_opt(proto.secs, proto.frac)
            .ok_or_else(|| {
                TryFromProtoError::DateConversionError(format!(
                    "NaiveTime::from_num_seconds_from_midnight_opt({},{}) failed",
                    proto.secs, proto.frac
                ))
            })?;

        Ok(NaiveDateTime::new(date, time))
    }
}

impl RustType<ProtoNaiveDateTime> for DateTime<Utc> {
    fn into_proto(&self) -> ProtoNaiveDateTime {
        self.naive_utc().into_proto()
    }

    fn from_proto(proto: ProtoNaiveDateTime) -> Result<Self, TryFromProtoError> {
        Ok(DateTime::from_naive_utc_and_offset(
            NaiveDateTime::from_proto(proto)?,
            Utc,
        ))
    }
}

pub fn any_naive_date() -> impl Strategy<Value = NaiveDate> {
    (0..1000000).prop_map(|d| NaiveDate::from_num_days_from_ce_opt(d).unwrap())
}

pub fn any_naive_datetime() -> impl Strategy<Value = NaiveDateTime> {
    (0..(NaiveDateTime::MAX.nanosecond() - NaiveDateTime::MIN.nanosecond()))
        .prop_map(|x| NaiveDateTime::MIN + Duration::nanoseconds(i64::from(x)))
}

pub fn any_datetime() -> impl Strategy<Value = DateTime<Utc>> {
    any_naive_datetime().prop_map(|x| DateTime::from_naive_utc_and_offset(x, Utc))
}

pub fn any_fixed_offset() -> impl Strategy<Value = FixedOffset> {
    (-86_399..86_400).prop_map(|o| FixedOffset::east_opt(o).unwrap())
}

pub fn any_timezone() -> impl Strategy<Value = Tz> {
    (0..TZ_VARIANTS.len()).prop_map(|idx| *TZ_VARIANTS.get(idx).unwrap())
}

#[cfg(test)]
mod tests {
    use crate::protobuf_roundtrip;
    use mz_ore::assert_ok;
    use proptest::prelude::*;

    use super::*;

    proptest! {
        #![proptest_config(ProptestConfig::with_cases(4096))]

        #[mz_ore::test]
        #[cfg_attr(miri, ignore)] // too slow
        fn naive_date_protobuf_roundtrip(expect in any_naive_date() ) {
            let actual = protobuf_roundtrip::<_, ProtoNaiveDate>(&expect);
            assert_ok!(actual);
            assert_eq!(actual.unwrap(), expect);
        }

        #[mz_ore::test]
        #[cfg_attr(miri, ignore)] // too slow
        fn naive_date_time_protobuf_roundtrip(expect in any_naive_datetime() ) {
            let actual = protobuf_roundtrip::<_, ProtoNaiveDateTime>(&expect);
            assert_ok!(actual);
            assert_eq!(actual.unwrap(), expect);
        }

        #[mz_ore::test]
        #[cfg_attr(miri, ignore)] // too slow
        fn date_time_protobuf_roundtrip(expect in any_datetime() ) {
            let actual = protobuf_roundtrip::<_, ProtoNaiveDateTime>(&expect);
            assert_ok!(actual);
            assert_eq!(actual.unwrap(), expect);
        }
    }
}
