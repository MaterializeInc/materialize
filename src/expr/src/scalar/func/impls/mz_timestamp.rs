// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use chrono::{DateTime, NaiveDateTime, Utc};

use mz_ore::result::ResultExt;
use mz_repr::{
    adt::{numeric::Numeric, timestamp::CheckedTimestamp},
    strconv, Timestamp,
};

use crate::EvalError;

// Conversions to MzTimestamp, and a single conversion from MzTimestamp to
// String. In general we want to make MzTimestamp a more opaque type, so we
// easily support casting things to it but not from it.

sqlfunc!(
    #[sqlname = "mz_timestamp_to_text"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastStringToMzTimestamp)]
    fn cast_mz_timestamp_to_string(a: Timestamp) -> String {
        let mut buf = String::new();
        strconv::format_mz_timestamp(&mut buf, a);
        buf
    }
);

sqlfunc!(
    #[sqlname = "text_to_mz_timestamp"]
    #[preserves_uniqueness = false]
    #[right_inverse = to_unary!(super::CastMzTimestampToString)]
    fn cast_string_to_mz_timestamp(a: String) -> Result<Timestamp, EvalError> {
        strconv::parse_mz_timestamp(&a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "numeric_to_mz_timestamp"]
    #[preserves_uniqueness = true]
    fn cast_numeric_to_mz_timestamp(a: Numeric) -> Result<Timestamp, EvalError> {
        // The try_into will error if the conversion is lossy (out of range or fractional).
        a.try_into().map_err(|_| EvalError::MzTimestampOutOfRange)
    }
);

sqlfunc!(
    #[sqlname = "uint8_to_mz_timestamp"]
    #[preserves_uniqueness = true]
    fn cast_uint64_to_mz_timestamp(a: u64) -> Timestamp {
        a.into()
    }
);

sqlfunc!(
    #[sqlname = "uint4_to_mz_timestamp"]
    #[preserves_uniqueness = true]
    fn cast_uint32_to_mz_timestamp(a: u32) -> Timestamp {
        u64::from(a).into()
    }
);

sqlfunc!(
    #[sqlname = "bigint_to_mz_timestamp"]
    #[preserves_uniqueness = true]
    fn cast_int64_to_mz_timestamp(a: i64) -> Result<Timestamp, EvalError> {
        a.try_into().map_err(|_| EvalError::MzTimestampOutOfRange)
    }
);

sqlfunc!(
    #[sqlname = "integer_to_mz_timestamp"]
    #[preserves_uniqueness = true]
    fn cast_int32_to_mz_timestamp(a: i32) -> Result<Timestamp, EvalError> {
        i64::from(a)
            .try_into()
            .map_err(|_| EvalError::MzTimestampOutOfRange)
    }
);

sqlfunc!(
    #[sqlname = "timestamp_tz_to_mz_timestamp"]
    fn cast_timestamp_tz_to_mz_timestamp(
        a: CheckedTimestamp<DateTime<Utc>>,
    ) -> Result<Timestamp, EvalError> {
        a.timestamp_millis()
            .try_into()
            .map_err(|_| EvalError::MzTimestampOutOfRange)
    }
);

sqlfunc!(
    #[sqlname = "timestamp_to_mz_timestamp"]
    fn cast_timestamp_to_mz_timestamp(
        a: CheckedTimestamp<NaiveDateTime>,
    ) -> Result<Timestamp, EvalError> {
        a.timestamp_millis()
            .try_into()
            .map_err(|_| EvalError::MzTimestampOutOfRange)
    }
);

sqlfunc!(
    #[sqlname = "step_mz_timestamp"]
    #[preserves_uniqueness = true]
    fn step_mz_timestamp(a: Timestamp) -> Result<Timestamp, EvalError> {
        a.checked_add(1).ok_or(EvalError::MzTimestampStepOverflow)
    }
);
