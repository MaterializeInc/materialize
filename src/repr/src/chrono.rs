// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Custom [`proptest::strategy::Strategy`] implementations and Protobuf types
//! for the [`chrono`] fields used in the codebase.
//!
//! See the [`proptest`] docs[^1] for an example.
//!
//! [^1]: <https://altsysrq.github.io/proptest-book/proptest-derive/modifiers.html#strategy>

use std::str::FromStr;

use chrono::{
    DateTime, Datelike, Duration, FixedOffset, NaiveDate, NaiveDateTime, NaiveTime, Timelike, Utc,
};
use chrono_tz::{Tz, TZ_VARIANTS};
use proptest::prelude::Strategy;

use mz_proto::{RustType, TryFromProtoError};

include!(concat!(env!("OUT_DIR"), "/mz_repr.chrono.rs"));

impl RustType<ProtoNaiveDate> for NaiveDate {
    fn into_proto(&self) -> ProtoNaiveDate {
        ProtoNaiveDate {
            year: self.year(),
            ordinal: self.ordinal(),
        }
    }

    fn from_proto(proto: ProtoNaiveDate) -> Result<Self, TryFromProtoError> {
        NaiveDate::from_yo_opt(proto.year, proto.ordinal).ok_or_else(|| {
            TryFromProtoError::DateConversionError(format!(
                "NaiveDate::from_yo_opt({},{}) failed",
                proto.year, proto.ordinal
            ))
        })
    }
}

impl RustType<ProtoNaiveTime> for NaiveTime {
    fn into_proto(&self) -> ProtoNaiveTime {
        ProtoNaiveTime {
            secs: self.num_seconds_from_midnight(),
            frac: self.nanosecond(),
        }
    }

    fn from_proto(proto: ProtoNaiveTime) -> Result<Self, TryFromProtoError> {
        NaiveTime::from_num_seconds_from_midnight_opt(proto.secs, proto.frac).ok_or_else(|| {
            TryFromProtoError::DateConversionError(format!(
                "NaiveTime::from_num_seconds_from_midnight_opt({},{}) failed",
                proto.secs, proto.frac
            ))
        })
    }
}

impl RustType<ProtoNaiveDateTime> for NaiveDateTime {
    fn into_proto(&self) -> ProtoNaiveDateTime {
        ProtoNaiveDateTime {
            year: self.year(),
            ordinal: self.ordinal(),
            secs: self.num_seconds_from_midnight(),
            frac: self.nanosecond(),
        }
    }

    fn from_proto(proto: ProtoNaiveDateTime) -> Result<Self, TryFromProtoError> {
        let date = NaiveDate::from_yo_opt(proto.year, proto.ordinal).ok_or_else(|| {
            TryFromProtoError::DateConversionError(format!(
                "NaiveDate::from_yo_opt({},{}) failed",
                proto.year, proto.ordinal
            ))
        })?;

        let time = NaiveTime::from_num_seconds_from_midnight_opt(proto.secs, proto.frac)
            .ok_or_else(|| {
                TryFromProtoError::DateConversionError(format!(
                    "NaiveTime::from_num_seconds_from_midnight_opt({},{}) failed",
                    proto.secs, proto.frac
                ))
            })?;

        Ok(NaiveDateTime::new(date, time))
    }
}

impl RustType<ProtoNaiveDateTime> for DateTime<Utc> {
    fn into_proto(&self) -> ProtoNaiveDateTime {
        self.naive_utc().into_proto()
    }

    fn from_proto(proto: ProtoNaiveDateTime) -> Result<Self, TryFromProtoError> {
        Ok(DateTime::from_utc(NaiveDateTime::from_proto(proto)?, Utc))
    }
}

impl RustType<ProtoFixedOffset> for FixedOffset {
    fn into_proto(&self) -> ProtoFixedOffset {
        ProtoFixedOffset {
            local_minus_utc: self.local_minus_utc(),
        }
    }

    fn from_proto(proto: ProtoFixedOffset) -> Result<Self, TryFromProtoError> {
        FixedOffset::east_opt(proto.local_minus_utc).ok_or_else(|| {
            TryFromProtoError::DateConversionError(format!(
                "FixedOffset::east_opt({}) failed.",
                proto.local_minus_utc
            ))
        })
    }
}

/// Encode a Tz as string representation. This is not the most space efficient solution, but
/// it is immune to changes in the chrono_tz (and is fully compatible with its public API).
impl RustType<ProtoTz> for chrono_tz::Tz {
    fn into_proto(&self) -> ProtoTz {
        ProtoTz {
            name: self.name().into(),
        }
    }

    fn from_proto(proto: ProtoTz) -> Result<Self, TryFromProtoError> {
        Tz::from_str(&proto.name).map_err(TryFromProtoError::DateConversionError)
    }
}

pub fn any_naive_date() -> impl Strategy<Value = NaiveDate> {
    (0..1000000).prop_map(|d| NaiveDate::from_num_days_from_ce_opt(d).unwrap())
}

pub fn any_naive_datetime() -> impl Strategy<Value = NaiveDateTime> {
    (0..(NaiveDateTime::MAX.nanosecond() - NaiveDateTime::MIN.nanosecond()))
        .prop_map(|x| NaiveDateTime::MIN + Duration::nanoseconds(i64::from(x)))
}

pub fn any_datetime() -> impl Strategy<Value = DateTime<Utc>> {
    any_naive_datetime().prop_map(|x| DateTime::from_utc(x, Utc))
}

pub fn any_fixed_offset() -> impl Strategy<Value = FixedOffset> {
    (-86_399..86_400).prop_map(|o| FixedOffset::east_opt(o).unwrap())
}

pub fn any_timezone() -> impl Strategy<Value = Tz> {
    (0..TZ_VARIANTS.len()).prop_map(|idx| *TZ_VARIANTS.get(idx).unwrap())
}

#[cfg(test)]
mod tests {
    use proptest::prelude::*;

    use super::*;
    use mz_proto::protobuf_roundtrip;

    proptest! {
        #![proptest_config(ProptestConfig::with_cases(4096))]

        #[test]
        fn naive_date_protobuf_roundtrip(expect in any_naive_date() ) {
            let actual = protobuf_roundtrip::<_, ProtoNaiveDate>(&expect);
            assert!(actual.is_ok());
            assert_eq!(actual.unwrap(), expect);
        }

        #[test]
        fn naive_date_time_protobuf_roundtrip(expect in any_naive_datetime() ) {
            let actual = protobuf_roundtrip::<_, ProtoNaiveDateTime>(&expect);
            assert!(actual.is_ok());
            assert_eq!(actual.unwrap(), expect);
        }

        #[test]
        fn date_time_protobuf_roundtrip(expect in any_datetime() ) {
            let actual = protobuf_roundtrip::<_, ProtoNaiveDateTime>(&expect);
            assert!(actual.is_ok());
            assert_eq!(actual.unwrap(), expect);
        }

        #[test]
        fn fixed_offset_protobuf_roundtrip(expect in any_fixed_offset() ) {
            let actual = protobuf_roundtrip::<_, ProtoFixedOffset>(&expect);
            assert!(actual.is_ok());
            assert_eq!(actual.unwrap(), expect);
        }
    }
}
