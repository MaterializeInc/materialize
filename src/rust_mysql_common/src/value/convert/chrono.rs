// Copyright (c) 2021 Anatoly Ikorsky
//
// Licensed under the Apache License, Version 2.0
// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT
// license <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. All files in the project carrying such notice may not be copied,
// modified, or distributed except according to those terms.

//! This module implements conversion from/to `Value` for `chrono` types.

#![cfg(feature = "chrono")]

use std::convert::TryFrom;

use chrono::{Datelike, NaiveDate, NaiveDateTime, NaiveTime, Timelike};

use crate::value::Value;

use super::{
    parse_mysql_datetime_string, parse_mysql_time_string, FromValue, FromValueError, ParseIr,
};

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl FromValue for NaiveDate {
    type Intermediate = ParseIr<NaiveDate>;
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl FromValue for NaiveTime {
    type Intermediate = ParseIr<NaiveTime>;
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl FromValue for NaiveDateTime {
    type Intermediate = ParseIr<NaiveDateTime>;
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl TryFrom<Value> for ParseIr<NaiveDateTime> {
    type Error = FromValueError;

    fn try_from(v: Value) -> Result<Self, Self::Error> {
        match v {
            Value::Date(year, month, day, hour, minute, second, micros) => {
                let date = NaiveDate::from_ymd_opt(year.into(), month.into(), day.into());
                let time = NaiveTime::from_hms_micro_opt(
                    hour.into(),
                    minute.into(),
                    second.into(),
                    micros,
                );
                if let Some((date, time)) = date.zip(time) {
                    Ok(ParseIr(NaiveDateTime::new(date, time), v))
                } else {
                    Err(FromValueError(v))
                }
            }
            Value::Bytes(ref bytes) => {
                if let Some((year, month, day, hour, minute, second, micros)) =
                    parse_mysql_datetime_string(bytes)
                {
                    let date = NaiveDate::from_ymd_opt(year as i32, month, day);
                    let time = NaiveTime::from_hms_micro_opt(hour, minute, second, micros);
                    if let Some((date, time)) = date.zip(time) {
                        Ok(ParseIr(NaiveDateTime::new(date, time), v))
                    } else {
                        Err(FromValueError(v))
                    }
                } else {
                    Err(FromValueError(v))
                }
            }
            _ => Err(FromValueError(v)),
        }
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl TryFrom<Value> for ParseIr<NaiveDate> {
    type Error = FromValueError;

    fn try_from(v: Value) -> Result<Self, Self::Error> {
        let result = match v {
            Value::Date(year, month, day, hour, minute, second, micros) => {
                let date = NaiveDate::from_ymd_opt(year.into(), month.into(), day.into());
                Ok((
                    date,
                    Value::Date(year, month, day, hour, minute, second, micros),
                ))
            }
            Value::Bytes(bytes) => {
                if let Some((y, m, d, _, _, _, _)) = parse_mysql_datetime_string(&*bytes) {
                    let date = NaiveDate::from_ymd_opt(y as i32, m, d);
                    Ok((date, Value::Bytes(bytes)))
                } else {
                    Err(FromValueError(Value::Bytes(bytes)))
                }
            }
            v => Err(FromValueError(v)),
        };

        let (date, value) = result?;

        if let Some(output) = date {
            Ok(ParseIr(output, value))
        } else {
            Err(FromValueError(value))
        }
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl TryFrom<Value> for ParseIr<NaiveTime> {
    type Error = FromValueError;

    fn try_from(v: Value) -> Result<Self, Self::Error> {
        let result = match v {
            Value::Time(false, 0, h, m, s, u) => {
                let time = NaiveTime::from_hms_micro_opt(h.into(), m.into(), s.into(), u);
                Ok((time, Value::Time(false, 0, h, m, s, u)))
            }
            Value::Bytes(bytes) => {
                if let Some((false, h, m, s, u)) = parse_mysql_time_string(&*bytes) {
                    let time = NaiveTime::from_hms_micro_opt(h, m as u32, s as u32, u);
                    Ok((time, Value::Bytes(bytes)))
                } else {
                    Err(FromValueError(Value::Bytes(bytes)))
                }
            }
            v => Err(FromValueError(v)),
        };

        let (time, value) = result?;

        if let Some(output) = time {
            Ok(ParseIr(output, value))
        } else {
            Err(FromValueError(value))
        }
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<ParseIr<NaiveDateTime>> for NaiveDateTime {
    fn from(value: ParseIr<NaiveDateTime>) -> Self {
        value.commit()
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<ParseIr<NaiveDate>> for NaiveDate {
    fn from(value: ParseIr<NaiveDate>) -> Self {
        value.commit()
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<ParseIr<NaiveTime>> for NaiveTime {
    fn from(value: ParseIr<NaiveTime>) -> Self {
        value.commit()
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<ParseIr<NaiveDateTime>> for Value {
    fn from(value: ParseIr<NaiveDateTime>) -> Self {
        value.rollback()
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<ParseIr<NaiveDate>> for Value {
    fn from(value: ParseIr<NaiveDate>) -> Self {
        value.rollback()
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<ParseIr<NaiveTime>> for Value {
    fn from(value: ParseIr<NaiveTime>) -> Self {
        value.rollback()
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<NaiveDateTime> for Value {
    fn from(x: NaiveDateTime) -> Value {
        if 1000 > x.year() || x.year() > 9999 {
            panic!("Year `{}` not in supported range [1000, 9999]", x.year())
        }
        Value::Date(
            x.year() as u16,
            x.month() as u8,
            x.day() as u8,
            x.hour() as u8,
            x.minute() as u8,
            x.second() as u8,
            x.nanosecond() / 1000,
        )
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<NaiveDate> for Value {
    fn from(x: NaiveDate) -> Value {
        if 1000 > x.year() || x.year() > 9999 {
            panic!("Year `{}` not in supported range [1000, 9999]", x.year())
        }
        Value::Date(x.year() as u16, x.month() as u8, x.day() as u8, 0, 0, 0, 0)
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "chrono")))]
impl From<NaiveTime> for Value {
    fn from(x: NaiveTime) -> Value {
        Value::Time(
            false,
            0,
            x.hour() as u8,
            x.minute() as u8,
            x.second() as u8,
            x.nanosecond() / 1000,
        )
    }
}
