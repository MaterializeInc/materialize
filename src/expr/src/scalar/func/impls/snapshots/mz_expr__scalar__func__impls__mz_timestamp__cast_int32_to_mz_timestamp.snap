---
source: src/expr/src/scalar/func/impls/mz_timestamp.rs
expression: "#[sqlfunc(\n    sqlname = \"integer_to_mz_timestamp\",\n    preserves_uniqueness = true,\n    inverse = None,\n    is_monotone = true,\n)]\n#[allow(clippy::extra_unused_lifetimes)]\npub fn cast_int32_to_mz_timestamp<'a>(a: i32) -> Result<Timestamp, EvalError> {\n    {\n        i64::from(a)\n            .try_into()\n            .map_err(|_| EvalError::MzTimestampOutOfRange(a.to_string().into()))\n    }\n}\n"
---
#[derive(
    proptest_derive::Arbitrary,
    Ord,
    PartialOrd,
    Clone,
    Debug,
    Eq,
    PartialEq,
    serde::Serialize,
    serde::Deserialize,
    Hash,
    mz_lowertest::MzReflect
)]
pub struct CastInt32ToMzTimestamp;
impl<'a> crate::func::EagerUnaryFunc<'a> for CastInt32ToMzTimestamp {
    type Input = i32;
    type Output = Result<Timestamp, EvalError>;
    fn call(&self, a: Self::Input) -> Self::Output {
        cast_int32_to_mz_timestamp(a)
    }
    fn output_type(&self, input_type: mz_repr::ColumnType) -> mz_repr::ColumnType {
        use mz_repr::AsColumnType;
        let output = Self::Output::as_column_type();
        let propagates_nulls = crate::func::EagerUnaryFunc::propagates_nulls(self);
        let nullable = output.nullable;
        output.nullable(nullable || (propagates_nulls && input_type.nullable))
    }
    fn inverse(&self) -> Option<crate::UnaryFunc> {
        None
    }
    fn is_monotone(&self) -> bool {
        true
    }
    fn preserves_uniqueness(&self) -> bool {
        true
    }
}
impl std::fmt::Display for CastInt32ToMzTimestamp {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        f.write_str("integer_to_mz_timestamp")
    }
}
#[allow(clippy::extra_unused_lifetimes)]
pub fn cast_int32_to_mz_timestamp<'a>(a: i32) -> Result<Timestamp, EvalError> {
    {
        i64::from(a)
            .try_into()
            .map_err(|_| EvalError::MzTimestampOutOfRange(a.to_string().into()))
    }
}
